### DISCUSSION TITLE
writting checkpoint for module

### DISCUSSION THREAD

**Discussion by GBrics** [Xilinx/RapidWright#369] [Q&A] [answered]
Hi again, 

This question follows my [previous question](https://github.com/Xilinx/RapidWright/discussions/363) on the encryption of checkpoints generated with Vivado 2020.2.

Now the previous issue was resolved, I'm faced to another problem:

When I write the checkpoint file for one of my modules, the generated DCP does not contain the routing information (except the intra-slice routing). All cells are properly placed, but the nets are not routed.

I tried a few things like creating a PBLOCK and assigning the module to it with CONTAIN_ROUTING and EXCLUDE_PLACEMENT set, applying DONT_TOUCH attribute on the module etc., but it didn't solve my problem.

The only solution I found for now is to create the module separately using the hierarchical design methodology (UG905) and then integrate the generated DCP as module in my design. 

Is there any way to export the module directly from the vivado project ? 
Is there any special consideration regarding the nets shared between the module and the rest of the design (clock signal) ? I would 
like to have a checkpoint file that includes the nodes of the clock net (at least starting from the BUFG).

Thank you in advance for you help !

Gbrics

**Comment by clavin-xlnx** [comment, marked as answer]
Hi Gbrics,

Are you calling `write_checkpoint -cell`?  That method is know to drop routing if the design is not configured in a specific way.  In order to preserve routing in a `write_checkpoint -cell` call, two things must be true:

1) `HD.RECONFIGURABLE` must be set to `1` on the cell being written out
2) In order for 1 to be possible, the design cannot be out-of-context

In Vivado, try setting the property by running the following Tcl command:
```
set_property HD.RECONFIGURABLE 1 [get_cells <your_cell_name>]
```
If this succeeds, then you should be able to write out the cell with `write_checkpoint -cell`.  If it fails, you'll want to change the DCP from out-of-context to in-context.  One way to do this is to use RapidWright.  Read in the DCP, change the setting and then write it back out:
```
Design design = Design.readCheckpoint("your_design.dcp");
design.setDesignOutOfContext(false);
design.writeCheckpoint("your_design_in_context.dcp");
```
Then you can load the resulting DCP back into Vivado and try the `set_property` command again.

Hope that helps.

Chris

**Comment by GBrics** [comment]
Hi Chris, 

Indeed, I'm using the ``write_checkpoint -cell`` command. 
Setting the property ``HD.RECONFIGURABLE`` solved my problem, all the routing information is now conserved in the checkpoint!

Thank you again for your precious help.

GBrics
